repos/system_upgrade/common/configs/rpmtransactionconfigtaskscollector.py

```diff
@@ -0,0 +1,44 @@
+from leapp.models.fields import List
+from leapp.models.configs import Config
+
+
+### Nested containers?
+### Duplication of default value in type_ and Config.  If we eliminate that, we need to extract default from the type_ for the documentation.
+### We probably want to allow dicts in Config.  But IIRC, dicts were specifically excluded for model fields.  Do we need something that restricts where fields are valid?
+### Another thing we might want is must_be_string (no conversion).  This can be helpful for things like file modes where 644 as a number would be wrong but "644" and "0644" can be inerpretted correctly
+class Transaction_ToInstall(Config):
+    section = "transaction"
+    name = "to_install"
+    type_ = fields.List(fields.String(), default=[])
+    default = []
+    description = """
+        List of packages to be added to the upgrade transaction.
+        Signed packages which are already installed will be skipped.
+    """
+
+
+class Transaction_ToKeep(Config):
+    section = "transaction"
+    name = "to_keep"
+    type_ = fields.List(fields.String(), default=[
+        "leapp",
+        "python2-leapp",
+        "python3-leapp",
+        "leapp-repository",
+        "snactor",
+    ])
+    description = """
+        List of packages to be kept in the upgrade transaction.
+    """
```

Comment:
Just trying to think why it cannot be same as with models? I mean:
```suggestion
class Transaction_ToKeep(Config):
    section = "transaction"
    to_keep = fields.List(fields.String(), default=[
        "leapp",
        "python2-leapp",
        "python3-leapp",
        "leapp-repository",
        "snactor",
    ])
    """
        List of packages to be kept in the upgrade transaction.
    """
```

Reply:
Since we've settled on yaml as the actual file system format, I think that loading more information into fields will be fine (default and type_).  to_keep and the docstring treatment were things I didn't know about.  I think they'll be fine too.

One ramification of relying on fields (I think in any incarnation) is that we should allow mapping values.  I heard that those were specifically excluded from the present fields because they were considered too complex and unnecessary for Models.  I do not think those are true for config values.  Yaml has dictionaries as basic data types and in cases where you want to group record-style information, yaml mappings make the most sense.
``` yaml
hosts:
    balancer.example.com: 192.168.100.2
    git.example.com: 192.168.100.3
```

I don't know if we'll want to keep field.Mapping (or field.Dict)  from being used in models or not care now that we have a need for it for a different purpose.

Reply:
@abadger  I was thinking about that during weekend, as I think I misunderstood you originally after reading this your comment and the example. I thought that this problem is addressed by [fields.JSON](https://github.com/oamg/leapp/blob/master/leapp/models/fields/__init__.py#L509) which we can allow for configs to not be so strict. However, thinking about that, I think you provided totally valid point about having something like a `Map`/`Mapping`/`Dict` field, which could have variable keys (like hosts, where we do not know number of such items, nor specific keys itself). But we could on the other hand still define what will be values like in case of other fields. So the use for the example above, I could imagine to allow something like:

```
class MyConfig(Config):
    hosts = fields.Mapping(fields.String)  # or e.g. fields.Mapping(value_type=fields.String)
```
Where we can define what will be values, hence the key is expected to be always string (no sure whether we would need anything else for keys than string). For the completness, I can imagine we could resolve it also differently, but usually it would lead to some magical tweaks around serialisation/deserialisation, or little bit worse UX, e.g. to require user to write config like:

```
hosts:
    - hostname: balancer.example.com
    - ip: 192.168.100.3
```

Which is doable and still kind of reasonable, but thinking about the trade-off, it make sense to me to be less strict here.

@abadger  Do I understand correctly your idea?

---
Just to provide context why we usually do not allow [fields.JSON](https://github.com/oamg/leapp/blob/master/leapp/models/fields/__init__.py#L509) in Models inside our official repositories. It's because the JSON field cannot be verified for the content. From the point, we can verify the structure is JSON, but we cannot verify it contains particular fields / content inside, so the structure can vary and we couldn't guarantee that part of this structure will be always a specific key/item (due to bugs, or changes in actor that produce a particular msg, so some other actors - e.g. custom ones - could start to be broken without clean visibility of the problem). Having strict structure in models, provides clear definition of the content and people can easily see it.